File: leetcode/minStack.js
/**
 * initialize your data structure here.
 */
var Node = function (value, min, next = null) {
	this.value = value;
	this.min = Math.min(value, min);
	this.next = next;
};

var MinStack = function () {
	this.head = null;
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function (x) {
	if (this.head) this.head = new Node(x, this.head.min, this.head);
	else this.head = new Node(x, x);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
	this.head = this.head.next;
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
	return this.head.value;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
	return this.head.min;
};

/*------------------------------------------------------------------------------------------*/
class MinStack {
	constructor() {
		this.stack = [];
		this.min = Infinity;
	}
	push(x) {
		if (this.stack.length === 0) {
			this.stack.push(0);
			this.min = x;
		} else {
			this.stack.push(x - min);
			if (x < min) this.min = x;
		}
	}
	pop() {
		if (this.stack.length > 0) {
			const top = this.stack.pop();
			if (top < 0) min = min - top;
		}
	}
	top() {
		const top = this.stack[this.stack.length - 1];
		if (top > 0) return top + this.min;
		return this.min;
	}
	getMin() {
		return this.min;
	}
}
/*------------------------------------------------------------------------------------------*/
class MinStack {
	constructor() {
		this.stack = [];
		this.min = Infinity;
	}
	push(x) {
		if (x <= this.min) {
			this.stack.push(this.min);
			this.min = x;
		}
		this.stack.push(x);
	}
	pop() {
		if (this.stack.pop() === this.min) {
			this.min = this.stack.pop();
		}
	}
	top() {
		return this.stack[this.stack.length - 1];
	}
	getMin() {
		return this.min;
	}
}
/*------------------------------------------------------------------------------------------*/
class MinStack1 {
	constructor() {
		this.stack = [];
	}
	push(x) {
		let min = this.getMin();

		if (min === null || (min > x)) {
			min = x;
		}
		this.stack.push([x, min]);
	}
	pop() {
		this.stack.pop();
	}
	top() {
		return this.stack.length
			? this.stack[this.stack.length - 1][0]
			: null;
	}
	getMin() {
		return this.stack.length
			? this.stack[this.stack.length - 1][1]
			: null;
	}
}

const stack = new MinStack1();
stack.push(0);
stack.push(1)
stack.push(0);
console.log(stack)
stack.getMin();
stack.pop();
stack.getMin();

File: leetcode/MinStack.py
class MinStack:

    def __init__(self):
        self.s1 = []
        self.s2 = []

    def push(self, val: int) -> None:
        self.s1.push(val)
        if self.s2 or val <= self.s2[-1]:
            self.s2.push(val)

    def pop(self) -> None:
        if self.top() == self.getMin():
            self.s2.pop()
        self.s1.pop()

    def top(self) -> int:
        return self.s1[-1]

    def getMin(self) -> int:
        return self.s2[-1]
        # Your MinStack object will be instantiated and called as such:
        # obj = MinStack()
        # obj.push(val)
        # obj.pop()
        # param_3 = obj.top()
        # param_4 = obj.getMin()


class MinStack1Stack:

    def __init__(self):
        self.s1 = []
        self.min = float('inf')

    def push(self, val: int) -> None:
        if val <= self.getMin():
            self.s1.append(self.getMin())
            self.min = val
        self.s1.append(val)

    def pop(self) -> None:
        popped = self.top()
        self.s1.pop()
        if popped == self.getMin():
            self.min = self.top()
            self.s1.pop()

    def top(self) -> int:
        return self.s1[-1]

    def getMin(self) -> int:
        return self.min
        # Your MinStack object will be instantiated and called as such:
        # obj = MinStack()
        # obj.push(val)
        # obj.pop()
        # param_3 = obj.top()
        # param_4 = obj.getMin()
